### ISSUE TITLE
Bug: circular import between shared/ReactSharedInternals.js and react/src/React.js

### ISSUE THREAD

**Issue by bpstrngr** [facebook/react#31126]
I'm trying to build React packages from source, 
by passing "packages/react/src/React.js" as the entry point to Rollup and transpiling the modules to ESM with flow-parser. 

Several modules in the `react` package import `shared/ReactSharedInternals` (directly or indirectly, eg. https://github.com/facebook/react/blob/v18.3.1/packages/react/src/ReactServerContext.js), but `shared/ReactSharedInternals` actually circularly imports "react" itself to access its `react/src/ReactSharedInternals.js` exposed as `React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED` (https://github.com/facebook/react/blob/v18.3.1/packages/shared/ReactSharedInternals.js#L13), resulting in this error: 
```
ReferenceError: Cannot access '__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED' before initialization
    at file:///home/blik/facebook_2013_react/0/packages/shared/ReactSharedInternals.js:2:36
```
or as Rollup reports: 
```
Circular dependency: facebook_2013_react/0/packages/react/src/React.js
 -> facebook_2013_react/0/packages/react/src/ReactElementValidator.js
 -> facebook_2013_react/0/packages/shared/checkPropTypes.js
 -> facebook_2013_react/0/packages/shared/ReactComponentStackFrame.js
 -> facebook_2013_react/0/packages/shared/ReactSharedInternals.js
 -> facebook_2013_react/0/packages/react/src/React.js
```

For now, I have redirected imports of `shared/ReactSharedInternals` to `react/src/ReactSharedInternals.js`, but since the "shared" package is used in other packages beside `react`, i should bundle it separately (possibly one by one because they have default exports), in which case i would be duplicating React by it being default-imported for this one property access, leaving no space for even any treeshaking: 
```js
  "facebook_2013_react.js":{"https://github.com/facebook/react/packages/react packages/shared":{"v18.3.1":["packages/react/src/React.js",{
    "syntax":"flow",
    "edit":{
      "__DEV__":"false",
      "packages/react/src/ReactServerContext.js":{"shared/ReactSharedInternals":"./ReactSharedInternals.js"},
      "packages/react":{"from 'shared/":"from '../../shared/"},
      "packages/shared":{"shared/ReactSharedInternals":"../react/src/ReactSharedInternals.js","from 'shared/":"from './","__EXPERIMENTAL__":"false","__PROFILE__":"false"}
    },
    "alias":{"shared/ReactSharedInternals.js":{"react":"./facebook_2013_react/0/packages/react/src/React.js"}},
    "comment":"shared/ReactSharedInternals circularly imports React, which imports it back indirectly. redirected those imports to the original react/src/ReactSharedInternals..."
  }]}},
//"facebook_2013_react_shared.js":{
//  "https://github.com/facebook/react/packages/shared":{"v18.3.1":[
//    "packages/shared/*.js",{"syntax":"flow","alias":{
//      "react":"./facebook_2013_react/0/packages/react/src/React.js"
//    },"comment":"default exports will be lost, react bundled unnecessarily"}
//  ]}
//}
```

- Why isn't `react/src/ReactSharedInternals.js` located in the "shared" package directly to resolve this circular import? 

- How does the official build process access `React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED` in `shared/ReactSharedInternals.js` given the "react" package's circular import graph? 

- How does it share the "shared" modules across packages if the "shared" package can't be bundled as a shared module (due to default exports and a circular import), only be re-included in each package's bundle? 

React version: 18.3.1

## Steps To Reproduce

1. Pass react/src/React.js to Rollup with a plugin to transpile flow 
```js
 import("./rollup.js").then(({rollup})=>rollup(
 {input:"packages/react/src/React.js"
 ,plugins:
[{transform:compose(flow,prunetypes,astring)
 ,resolveId:(source,client)=>source.replace(/^shared/
,shared=>(/react\/src/.test(client)?"../../":"./")+shared)
 }
]}));
```
2. see errors of circular import between `packages/shared/ReactSharedInternals.js -> packages/react/src/React.js`

## The current behavior

Circular import

## The expected behavior

A "shared" module trivial to bundle, and no circular imports from it to peer packages.

**Comment by markerikson** [comment]
I'm a bit confused on the overall context. Are you trying to put together some kind of custom build setup that is separate than the existing React package build pipeline?  If so, why?

**Comment by bpstrngr** [comment]
Yes, well I'm not using npm, but building dependencies from source so that I can have the ability to maintain/customize them instead of patching arbitrary build outputs. Not necessarily React in particular, but being able to build them is my bare minimum for any dependency. 
Either way I did try to rely on the existing pipeline, but they're (expectably) rather complicated (eg. haven't spotted where package cross-specifiers are replaced in imports) and documentations don't extend beyond a user perspective of "npm install react react-dom". It was easier to figure out from looking at the source files, where I noticed this particular circular import and the multiple default exports in "shared" package which prevent bundling eg. "react" and "react-dom" (etc.) separately (either you bundle them all together, or you re-include "shared" modules in each, neither of which seem possible without redirecting from shared/ReactSharedInternals' un-treeshakable circular namespace import of "react" to react/src/ReactSharedInternals directly). 
The question is simply whether the "shared" package is intended to be modular despite its multiple peripheral defaultexports, and even if not, 
how does the official build pipeline resolve this seemingly even unnecessary circular import (import("react") -> shared -> react -> shared -> ⛔ can't access react/src/sharedinternals on react before exported by it) for the distributed packages? 
If neither is known, then I'll just continue with my redirect workaround and "shared" redundancies.

**Comment by markerikson** [comment]
I don't have a specific answer here, but I agree that the React build pipeline is _incredibly_ complicated.  Given that, my own advice would be to rely on some modification of that existing pipeline as a running fork, and not try to create your own.

What actual build changes or lib changes are you trying to accomplish?

**Comment by bpstrngr** [comment]
that's alright, thank you! 
2 of them: 
- no redundancy of "shared" modules if I want to bundle eg. "react" and "react-dom" (etc) separately, by "shared" being a truly modular package or else split into modules that only some peers share. the redundancy is not huge, but hard to tell the overlap/difference. 
- much more simply why is ReactSharedInternals located in "react" if it is re-exposed from "shared" to other packages, and "react" already imports from it just as well (or worse, because this particular cycle breaks bundling by the source's design).

**Comment by markerikson** [comment]
Heh, not quite what I was asking :)  Those are "why is React's module and build setup designed this way?", which I can't really help with.

I'm asking why you're even trying to customize React's behavior at all - what customizations are you trying to do that led you to try to build React in the first place?  And given that, why _aren't_ you just using the existing build setup as-is?

(unless I'm misunderstanding and "React's shared internals" _is_ related to the original reason why you're doing any of this at all?)

**Comment by bpstrngr** [comment]
I don't intend to change anything in React's functionality, just trying to gain control of its source, as part of becoming less dependent on and more conscious of npm's bottomless dependency graph of build outputs. I expect this to be possible, and it was, barring redundancies of "shared" "modules" and and edit for the package bundles. So the question is about the architecture, and these 2 issues are indeed what I would change, unless there is an answer to the "why" that I'm unaware of, and can't piece together from the build configs.

**Comment by markerikson** [comment]
Honestly, I'm pretty confused at what you're trying to accomplish, then :)

React's build system works as-is.

- How is building it from source yourself "being less dependent on NPM"?
- How is writing your own build setup for React going to improve the situation?
- If you're concerned about _installing_ React from NPM, why not download it once and vendor the pre-built artifacts?

Beyond that, given how complex React's internals are, and the way that the "shared" pieces actually get loaded and referenced in a page... I would be suspicious that a homegrown build setup of the same React source files would actually work correctly.

**Comment by bpstrngr** [comment]
- not using npm is as independent from npm as it gets. 
- writing my own build setup improves the situation of not being able to report on details of the current build pipeline (I successfully built with the total of ~3 simple edits above, plus flow-parser, astring and rollup without config. I'll see soon how it works to tell if you're right and there's more to it). 
- that would not prove the replicability/maintainability of the build to the user (me). 
If the integrity of my current bundles will seem corrupt ( I just got to remix's react-router for a sufficient test) I'll see what I can do further, but beyond a complexity I'll certainly cave to leave React as a humanly irreproducible source architecture and therefore fork or favor alternatives. At this point I only had these 2 simple observations about it. :)

**Comment by markerikson** [comment]
Yeah, I'll give up on this thread at this point.

I don't understand the point of trying to recreate a build pipeline that has been iterated on by dozens of engineers over 10+ years and is known to work correctly, especially when it's a build pipeline as complex as React's.  I can _vaguely_ understand choosing to build React from scratch yourself just so you can verify the output matches the source, but I don't understand why you feel you need to go about things this way, and I strongly suspect the builds you're generating aren't going to work properly.

**Comment by bpstrngr** [comment]
that's alright, thanks for your thoughts so far. there are marginal benefits to having modular bundles: I can reuse parts of React in other contexts (eg. react-reconciler, which given these 2 issues can't be bundled without React itself and any other peer dependency). I only filed a bug report because the source by design has this circular import, which would be unnecessary, however the build resolves it. I can always be wrong too of course. :)
